### PR TITLE
Lesson#12 Add OmniAuth Facebook

### PR DESCRIPTION
**Pull request by ValAntG**

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -15,7 +15,7 @@ gem 'remotipart', github: 'jbox-web/remotipart', tag: '1.5.0'
 gem 'responders'
 gem 'therubyracer'
 gem 'twitter-bootstrap-rails'
-gem 'virtus', '~> 1.0', '>= 1.0.5'
+# gem 'virtus', '~> 1.0', '>= 1.0.5'
```
Comment (svidlonix) [inline Gemfile]: Please delete unnecessary legacy code

File: app/models/user.rb [hunk 2]
```diff
@@ -1,8 +1,21 @@
...
-  # :confirmable, :lockable, :timeoutable, :trackable and :omniauthable
   devise :database_authenticatable, :registerable,
-         :recoverable, :rememberable, :validatable
+         :recoverable, :rememberable, :validatable, :omniauthable, omniauth_providers: [:facebook]
   has_many :questions
   has_many :answers
+  has_many :authorizations
+
+  def self.find_for_oauth(auth)
+    authorization = Authorization.where(provider: auth.provider, uid: auth.uid.to_s).first
```
Comment (svidlonix) [inline app/models/user.rb]: I think here better use `.find_by` instead of where

File: app/models/user.rb [hunk 3]
```diff
@@ -1,8 +1,21 @@
...
+    return authorization.user if authorization
+
+    email = auth.info[:email]
+    user = User.where(email: email).first
```
Comment (svidlonix) [inline app/models/user.rb]: I think here better use `.find_by` instead of where

File: spec/models/user_spec.rb [hunk 4]
```diff
@@ -5,4 +5,71 @@
   it { should have_many :answers }
   it { should validate_presence_of :email }
   it { should validate_presence_of :password }
+
+  describe '.find_for_oauth' do
+    let!(:user) { create(:user) }
```
Comment (svidlonix) [inline spec/models/user_spec.rb]: How is it possible rubocop don't warning you about `let!`?

File: spec/models/user_spec.rb [hunk 5]
```diff
@@ -5,4 +5,71 @@
...
+    let(:auth) { OmniAuth::AuthHash.new(provider: 'facebook', uid: '123456') }
+    let(:user_find_for_oauth) { described_class.find_for_oauth(auth) }
```
Comment (svidlonix) [inline spec/models/user_spec.rb]: Pretty uncleared whats happen here.

File: spec/models/user_spec.rb (hunk 4 above)
Comment (ValAntG) [reply spec/models/user_spec.rb]: No

Comment (svidlonix) [reply spec/models/user_spec.rb]: Please tune it in write a way to warn you.

File: config/credentials.yml.enc [hunk 6]
```diff
@@ -1 +1 @@
-NZBwq9bjgOQK3UcnU+YfExAAyPcOI5R2TjX5aaO1FWnYiROncJh+ztnfiYqFIsBz0Q1R3N8f7cmGqj1nxmYUEbgvGNh43ShqKj5FvgudK6sbOjM7W/l2jKJgDjpGgG7RLFOLf3RRyBkuVjnYWkftHQbur7ceRzJ1Xy+wnk5deHmRXAm0goYx3v8cX4UEUp2L+oLX1dqI8O/ss2PFYCC8NmpQB18Z3AMxfBmk9xdyY6CQW5YYzUJLFfy/leVxn0eJ/SCxcEdUYLlcw3u2D7sjTn1QHE49vHQ8J6HzZH1ZSAcBkX0PkxryWw1yifJyvon3ulvLQb8tOlCArE0v71ZwLNJdmB6E0knhecFfFOHYHZO+v7jsDFdr82ZbAtJp9SM/70uMAEmh8Gv4Xa44zdcxoDdFlPaji6GeEad5--0ugWqKp5QkeBPXQN--bT0Xiem7FRbnoFtU0qlbpg==
\ No newline at end of file
+/KanGLXNZi3YrwW4KNXWORH67F/Db1hJDEJuQN9CQJcQl1xm96wLTZFhw8bATw7//JeQR+lJ81ESh33pojjpNFm4vjzsMAMaHWkNxc3RjrEk9u+n4TcSHqWAdf7YPAvLGT+anuRo3CC9zTI1sZ1YW+tAnaq63RPKVkjpcLeHsIxbXXX5UjfHZRKEuZ+zLqfB7GweMw4HpW40lNvVWSEAXxxCP5ouz2cwlNULZEWiNDrJDjW/arbds7CuOClXE0Rc89SbEaiV/eIoB4rRA9IuJxj8FyTVBfTn4IaHhiLe+RDxUzQKex2V20TZZbRHGeN5FjXuaMZu2RNjkhixksK5R0EF8YSD60lh4b2AwPnAnmfzq42onYiGtGKBSNKPQ94fK+QicVNcK8iLbabkcwayNxF3qzKS5Bm8k/C7o8rKOROktENQFYoYn6X1pR0UZ61h1ydReSB5VbsbM4uyKkCDcvK0yfs7A+53K70KecL09oBYq05a5EQLcfyNnu7PlTBLOY0uUyrYPlPy+vK/z4ZyMYuHy2oGXWaQYfMHmzsyw7vko5uaqHJ8Oc9oD4IZWL9q2IW4gkso065CyIfe3s2a3u+vQnPF+NDOfmiwgepG4eHZky4tY1scYI2kgXg=--m4aamBCyhk2WMYZh--D1IYu/gkTyQRGtFeG1sOXg==
```
Comment (svidlonix) [inline config/credentials.yml.enc]: Please add last empty line.

Comment (ValAntG) [reply config/credentials.yml.enc]: This file itself is encrypted, I can’t add

**Pull request merged**